### ISSUE TITLE
What to do with tortured USFM

### ISSUE THREAD

**Issue by mvahowe** [mvh-solutions/nice-usfm-json#5]
Two real examples, received in the last week:
```
\id XXD - World English Bible
\h Bible reading plan
\toc1 Bible reading plan
\mt1 Daily Bible reading plan
\c 1
\cp January
\s January
\tr \tc1 date \tc2 Scripture
\tr \tc1 1 \tc2 \xt Genesis 1–2; Matthew 1:1–2:12; Psalm 1; Proverbs 1:1-6\xt*
```
I'd like to see chapter/verse checks on that.

```
\id XXF - World English Bible
\h Videos
\toc1 Videos
\toc2 Videos
\toc3 Videos
\mt2 Videos
\c 1
\cp Matthew
\s1 Gospel Films - Matthew (Lumo)
\s2 Visit for more resources: [www.ScriptureEarth.org/eng](https://www.scriptureearth.org/eng)
\b
\video VGF_ENGNIVP2DV_001
\b
```
Here we need to decide whether the b tag is appropriate, which, according to the USFM spec, depends on whether we think that a list of video filenames counts as poetry.

The semantics here are
- \c == arbitrary top-level division
- \cp == way to attach a label to that arbitrary top-level division

This is the big problem I see with all attempts to impose, retroactively, normalized semantics on USFM. These two examples are from a major, mature product from a major USFM stakeholder. Anything that stops this from working is a bug, not a feature. It would be nice if everyone adopted whatever the alternative to this turned out to be, but Agile, ETEN goals and general organisational inertia pretty much guarantee that things will work in the other direction, ie the technology that accepts all the existing weirdness without complaining is the technology that will be adopted.

Bottom line: not supporting this in Proskomma isn't really an option, and the same probably goes for all other tools used by more than one organisation.

**Comment by kavitharaju** [comment]
Its quite interesting to see people getting creative with USFM like this :rofl: 

With a grammar, we can say if it is syntactically correct or wrong, but could give absurd results if we open this in a Bible reading app or something. Even though `\c` is used I see `\v` is not used, which is a relief, that an app extracting scripture  (if it uses BCV, not paragraphs to do that) will not get any funny stuff from it.

**Comment by mvahowe** [comment]
What people generally want is either
- printed/rendered output that looks something like what they saw when they typed this into a well-known editor or (in this case)
- a data structure that works alongside "real" USFM to (in the first case) produce a lectionary (ie chunks of Scripture organised according to a church calendar)

**Comment by kavitharaju** [comment]
Can we generalize like that? I see USFM being used for scripture more often. And not only for rendering but also for data exchange. These kind of usages might be there, but is not the "general" trend, I think.

**Comment by mvahowe** [comment]
I'm talking about the tortured examples. No-one expects them to "just work" as Scripture, they expect them to "just work" as something else, and I would say "just work" in that case typically means one of the two options I listed.

**Comment by kavitharaju** [comment]
Then as long as its valid USFM, without `\video` and such tags, we can give a corresponding JSON representation for it, with our current approach of direct USX to JSON representation., may be.

**Comment by mvahowe** [comment]
I'm thinking that, somehow and at some level, things like `\video` or at least `\zvideo` need to Just Work. One challenge here is obviously that, in USFM, there's no way to tell whether `\video` is a paragraph or character tag. I'm wondering if we could add that kind of information in the headers. (That would still mean changing legacy tortured USFM but it could be done globally and/or automagically.)

**Comment by mvahowe** [comment]
eg
`\customtag video character`
(and maybe some other parsing information - just the above would be enough for me to work with)

**Comment by kavitharaju** [comment]
So you are expecting the USFM to come with that information? Then like you said the z-namespce or milestone, where it allows us to have custom names, will be the solution,right? And JSON representation will be supported out of the box as it is valid USFM.

Or detect invalid usages in USFM and represent that in the JSON with the extra info on it? IF that is the case keeping the information on `closing` could be an indication of character or para type tag.

**Comment by mvahowe** [comment]
Does the z-namespace tell me whether `\zvideo` is a paragraph style or a character style? I'm pretty sure that a well-known USFM editor blows up unless you explicitly provide that parsing information.

**Comment by mvahowe** [comment]
I don't think that `closing` ever clarifies anything since, according to the spec, the previous tag closes when you open a new tag. It's just convention that this is only typically used in places like footnotes.

**Comment by mvahowe** [comment]
My (thinking outloud) suggestion was
```
\id XXF - World English Bible
\customtag video character
\customtag banana paragraph
...
\h Videos
\mt2 Videos
\c 1
\video VGF_ENGNIVP2DV_001
```

**Comment by kavitharaju** [comment]
Oh, I see this is extending the USFM spec. I dont know if that is within our scope...

To answer your previous quetsion, znamespaces are milestones. Does that make them character type?

**Comment by mvahowe** [comment]
>  I dont know if that is within our scope...

Maybe you could point Martin at this... What I'm saying is that unless there's a way to handle this kind of thing that has previously Just Worked in the new USFM Spec, Proskomma won't be able to adopt the new USFM Spec, and that will be true of a lot of organisations with legacy data and workflows. There's no auto-migration fix here because everyone has literally rolled their own semantics one app or even one desktop at a time. Yes, that's a USFM spec-level issue. But, also, Proskomma won't be able to use any parser based on that spec if it makes handling this kind of data impossible.

**Comment by mvahowe** [comment]
Are 'znamespaces' anywhere in the spec? I know about z-prefixed tags, is that what we are talking about?

**Comment by kavitharaju** [comment]
It see it mentioned in [the milestones page](https://docs.usfm.bible/usfm-usx-docs/latest/ms/index.html#_standalone_milestones) of the new spec, with a link to it's own page saying extensions.

**Comment by kavitharaju** [comment]
May be your format can be 
```
\id XXF - World English Bible
\zcustomtag video character
\zcustomtag banana paragraph
...
\h Videos
\mt2 Videos
\c 1
\zvideo VGF_ENGNIVP2DV_001
```

**Comment by mvahowe** [comment]
I could live with that but it means changes throughout the document.

I've invited input from some people who actually use this type of USFM.

**Comment by mvahowe** [comment]
I think that
```\zvideo-s\* VGF_ENGNIVP2DV_001 \zvideo-e\*```
would work now in Proskomma (since the milestone syntax is unambiguous).

**Comment by mvahowe** [comment]
> \zcustomtag video character

The problem here is knowing that `\zcustomtag` is header metadata, for which you'd need another definition, which would have to be z-prefixed, which would require it's own definition, which... at some point you need the indirection to land on something with agreed semantics.

**Comment by kavitharaju** [comment]
The z-prefix itself says it is "privately defined" right? So why define it, as long as the  "private tools" understand it?

**Comment by mvahowe** [comment]
Because, eg, you might want to be able to edit the USFM with a standard editor, rather than having to write your own private editor.

**Comment by mvahowe** [comment]
I'm pretty sure that even z-milestones blow up Paratext because Paratext needs to know stuff about each tag it processes. Obviously if you write USFM in vi this isn't an issue, but most USFM power users use a USFM editor.

**Comment by mvahowe** [comment]
And the newer editors don't display USFM at all.

**Comment by mvahowe** [comment]
> I'm pretty sure that even z-milestones blow up Paratext because Paratext needs to know stuff about each tag it processes. Obviously if you write USFM in vi this isn't an issue, but most USFM power users use a USFM editor.

![image](https://user-images.githubusercontent.com/1576594/231743461-33462ae6-b74d-4a82-995d-199544e99de2.png)

**Comment by kavitharaju** [comment]
Did you try \zvideo​ ?
________________________________

**Comment by mvahowe** [comment]
This is an old copy of Paratext and I may be Doing It Wrong. But last time I asked, unfoldingWord USFM didn't work out of the box in Paratext because of the zaln milestones, and the solution was to manually configure Paratext to accept them. That's not a solution that makes for open data.

**Comment by mvahowe** [comment]
I chatted to Jeff Klassen about this and he shared his screen with a current version of Paratext. It doesn't crash like the screenshot above.

`\video` is treated as character-level markup.

Paratext adds a 3-letter abbreviation at the start of the \id line and capitalizes the rest of that line.

To get 'proper' handling of new tags, including z-prefixed milestones, you add them to the custom.sty file.

My point with all this wasn't that Paratext is broken. Paratext, like all other technology, is trying to navigate between expecting and working with spec-defined semantics and also handling all the weird stuff that people do. I'd like to have a conversation about how to do that.

**Comment by klassenjm** [comment]
The exception in Paratext may be about something else. I tested `\video` and some other undefined SFM markup in Paratext 8.0 and 9.2, and did not experience any crashes (saved, and moved between different visual modes - formatted, standard etc.). Paratext just shows unknown markup in red, and it does not know any of its properties -- so, for example, it does not know if \video is a new paragraph, or a character span.

Current -- in Paratext -- you need to add some basic definition for a marker to a `custom.sty` file -- things like whether it is paragraph or character, publishable, verse text, etc. Not all of it is required. The more you include, the better the checking and formatting tools can work. Mark - it looks like you're proposing a method to add a basic definition for user-generated `\z...` markup within the header of a USFM file (like what custom.sty is for Paratext).

**Comment by mvahowe** [comment]
> Mark - it looks like you're proposing a method to add a basic definition for user-generated \z... markup within the header of a USFM file (like what custom.sty is for Paratext).

Yes, the potential benefit being that the file then 'works' at some level in any compliant system. If that was considered likely to break things, we could consider workarounds like a REM-in-header convention for providing this information.

**Comment by mhosken** [comment]
First I have to remember that a demand that something must work is actually just a request. Currently \video is way outside the USFM spec and so what you are asking for is something beyond what is currently supported.

Yes, I think we do need some way for USFM to be more self describing with regard to custom markers. There is a long way to go on this and perhaps I should just write a proposal to the USFM/X committee for something. What I will say is that I will be surprised if markers not starting with z will be supported by such a mechanism.

BTW I would recommend your using a milestone for out of band information like this. So, perhaps something like:
```
\zvideo|VGF_ENGNIVP2DV_001\*
```

As to the grammar being more than the grammar. There is a plan that is only a mad idea at the moment, due to lack of capacity, to start adding schematron 'business rules' to ensure things like \v 2 can't come before \v 1. Although even that can be contentious, sigh.

**Comment by mvahowe** [comment]
> First I have to remember that a demand that something must work is actually just a request. Currently \video is way outside the USFM spec and so what you are asking for is something beyond what is currently supported.

It is, but a major ecosystem from your own organisation relies on it. This is what I keep coming back to. If your standard breaks major ecosystems, it will only be of academic interest, especially if there isn't a simple migration path.

> What I will say is that I will be surprised if markers not starting with z will be supported by such a mechanism.

Personally I can live with that. Out of interest, is there any norm for unknown attribute names? All the examples I've seen begin with `x-`. I can't immediately see that in the spec. If that is indeed the requirement, I'm going to suggest that we find ways to use the other 24 letters alphabet to mark up essentially the same kind of thing, in the interests of fairness to the letter j.

> BTW I would recommend your using a milestone for out of band information like this. So, perhaps something like:
> 
> ```
> \zvideo|VGF_ENGNIVP2DV_001\*
> ```

Yep, that was my recommendation, except that it doesn't currently work in Proskomma because I'd never seen a standalone milestone with an attribute. (I don't think there's an example in the spec.) May fix this next week...

> As to the grammar being more than the grammar. There is a plan that is only a mad idea at the moment, due to lack of capacity, to start adding schematron 'business rules' to ensure things like \v 2 can't come before \v 1. Although even that can be contentious, sigh.

It's contentious because a significant number of scholars reorder verses in some places. There's a notorious case in Zachariah where they move a non-integer number of verses (between 6b and 10 by memory). In other words, it's not resistance to change, it's connection to the messy realities of the domain.

**Comment by mvahowe** [comment]
> Currently \video is way outside the USFM spec

I did start writing a proposal for encoding video codecs in USFM a while ago. It wasn't entirely serious.

**Comment by mvahowe** [comment]
![zec4_versification](https://user-images.githubusercontent.com/1576594/234270900-74a5ff0f-c5db-45b3-bd14-69058d4c5f9f.png)

Enjoy... This is a major UBS translation.

**Comment by chrisvire** [comment]
Sorry that I am late to this conversation.  Mark asked me a while ago to join the conversation and I forgot ("one thing drives out another" - Barliman Butterbur).

The fact that the above example come from me (well actually Scripture App Builder), I feel like I should say something.  These two examples are extra-biblical content (a reading plan and a list of videos).  These can be added to SAB with DOCX, but they get converted to SFM (not USFM).  Richard needed to come up with something and it was only consumed by our internal parser, so didn't try very hard to get them to be USFM.

For our new PWA, we are using Proskomma as the USFM parser and we didn't anticipate this being an issue.  Thankfully, we have a processing step in between SAB exporting the content in SFM and the content being used by Proskomma.  In the case of the `\video VGF_ENGNIVP2DV_001` tag, I was able to covert these to `\zvideo-s |id="VGF_ENGNIVP2DV_001"\*\zvideo-e\*` and it is transparent to the user of the PWA.

In the case of the `\cp` tag, it is used to "label" the video "chapters" so that the chapter selector can have a name that shows up in the chapter dropdown in the SAB app instead of just a number.  Is this outside of the normal USFM?  Again, this is for being able to communicate metadata in the SFM file to the app so that it can affect the UI. 

With our processing step mentioned above, we would have the opportunity to inject extra markup into the SFM file or extra configuration into the Proskomma object to handle these non-USFM tags. Would that be similar to the data in the custom.sty files for Paratext?

**Comment by chrisvire** [comment]
Examples of `\cp` being using in the app:

![image](https://user-images.githubusercontent.com/517814/235171262-8c046c65-63c0-4a91-aa0b-952de5008c59.png)

![image](https://user-images.githubusercontent.com/517814/235171356-af7b47be-d69c-4676-91ca-7cb322df1e7b.png)

**Comment by alilland** [comment]
this is just a personal opinion as a software developer observing the conversations from outside the ecosystem, I did read the whole thread, but from a design standpoint it makes no sense to me that a markup language for bible translations would be skewed into other formats to support text that literally can be in any unpredictable format.

In my mind you strongly protect the bible translation syntax and dont provide flex beyond it, you provide a strong core that moves well from format to format, but you leave things like devotionals, or other generic content to be ported over to something like markdown, or another generic text format thats that more friendly to a "blog". Client applications that deviate from the reason the language exists need to fight private battles for how they want to support data. it should not be the bible translation markup language.

It should be a client by client choice for how to handle extra data outside of literal bible translations. Devotionals should not be taken into account for keeping data uniform. You will endlessly chase your tail and hinder a spec from being moved into use.
